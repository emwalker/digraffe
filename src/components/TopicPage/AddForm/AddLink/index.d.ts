import { RelayProp } from 'react-relay';
import { AddLink_viewer as ViewerType } from '__generated__/AddLink_viewer.graphql';
import { AddLink_topic as TopicType } from '__generated__/AddLink_topic.graphql';
declare type Props = {
    disabled?: boolean;
    relay: RelayProp;
    topic: TopicType;
    viewer: ViewerType;
};
declare const _default: import("react-relay").Container<Props>;
export default _default;
