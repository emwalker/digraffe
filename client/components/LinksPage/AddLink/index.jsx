// @flow
import React, { Component } from 'react'
import { Button, FormGroup, Input } from 'reactstrap'

import createLinkMutation from '../../../mutations/createLinkMutation'

type Props = {
  organization: {
    id: string,
    resourceId: string,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  state = {
    url: '',
  }

  onSubmit = () => {
    const {
      id: orgId,
      resourceId: organizationResourceId,
    } = this.props.organization
    createLinkMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationResourceId,
        url: this.state.url,
      },
    )
    this.setState({ url: '' })
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  render = () => (
    <div className="form-container">
      <FormGroup>
        <Input
          className="link-url test-link-url"
          placeholder="Link URL"
          value={this.state.url}
          onChange={this.updateUrl}
        />
        <Button size="sm" onClick={this.onSubmit}>Add</Button>
      </FormGroup>
    </div>
  )
}

export default AddLink
