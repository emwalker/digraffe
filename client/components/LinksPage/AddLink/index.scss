.form-container {
  text-align: right;
}

.link-url {
  display: inline-block;
  width: 75%;
  margin-right: 1em;
}
