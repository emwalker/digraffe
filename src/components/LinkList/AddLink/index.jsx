// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'
import Select from 'react-select'
import { path } from 'ramda'

import createLinkMutation from '../../../mutations/createLinkMutation'
import selectTopicMutation from '../../../mutations/selectTopicMutation'
import { liftNodes } from '../../../utils'

const selectedTopic = path(['selectedTopic', 'value'])

type Props = {
  organization: {
    id: string,
    resourceId: string,
    availableTopics: Array<{
      label: string,
      value: string,
    }>,
  },
  viewer: {
    selectedTopic: {
      label: string,
      value: string,
    },
  },
  relay: {
    environment: Object,
  },
}

type State = {
  selectedValue: string,
  url: string,
}

class AddLink extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.options = liftNodes(props.organization.availableTopics)
    this.state = {
      selectedValue: selectedTopic(props.viewer),
      url: '',
    }
  }

  onInputChange = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.setState(({ url }) => {
        this.addLink(url)
        return { url: '' }
      })
  }

  onSelect = (option) => {
    const selectedValue = option ? option.value : null
    this.setState({ selectedValue }, () => {
      selectTopicMutation(
        this.props.relay.environment,
        {
          organizationId: this.props.organization.id,
          topicId: selectedValue || '',
        },
      )
    })
  }

  addLink = (url) => {
    const {
      id: orgId,
      resourceId: organizationId,
    } = this.props.organization
    const value = this.state.selectedValue
    const topicIds = value ? [value] : []

    createLinkMutation(
      this.props.relay.environment,
      orgId,
      {
        organizationId,
        url,
        topicIds,
      },
    )
  }

  options: Object[]

  render = () => (
    <div>
      <dl className="form-group">
        <Select
          id="link-topic-id"
          name="link-topic"
          onChange={this.onSelect}
          options={this.options}
          placeholder="Select a topic"
          value={this.state.selectedValue}
        />
      </dl>
      <dl className="form-group">
        <input
          className="form-control test-link-url"
          onChange={this.onInputChange}
          onKeyPress={this.onKeyPress}
          placeholder="Link URL"
          value={this.state.url}
        />
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_viewer on User {
    selectedTopic {
      label: name
      value: resourceId
    }
  }

  fragment AddLink_organization on Organization {
    id
    resourceId

    availableTopics: topics(first: 100) {
      edges {
        node {
          label: name
          value: resourceId
        }
      }
    }
  }
`)
