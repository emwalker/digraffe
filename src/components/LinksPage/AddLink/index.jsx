// @flow
import React, { Component } from 'react'
import { FormGroup, Input } from 'reactstrap'
import { graphql, createFragmentContainer } from 'react-relay'
import Select from 'react-select'

import createLinkMutation from '../../../mutations/createLinkMutation'
import { liftNodes } from '../../../utils'

type Props = {
  organization: {
    id: string,
    resourceId: string,
    availableTopics: Array<{
      label: string,
      value: string,
    }>,
  },
  relay: {
    environment: Object,
  },
}

type State = {
  selectedOption: string,
  url: string,
}

class AddLink extends Component<Props, State> {
  constructor(props: Props) {
    super(props)
    this.options = liftNodes(props.organization.availableTopics)
  }

  state = {
    selectedOption: '',
    url: '',
  }

  onInputChange = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.setState(({ url }) => {
        this.addUrl(url)
        return { url: '' }
      })
  }

  onSelect = (selectedOption) => {
    this.setState({ selectedOption })
  }

  addUrl = (url) => {
    const {
      id: orgId,
      resourceId: organizationResourceId,
    } = this.props.organization
    createLinkMutation(
      this.props.relay.environment,
      orgId,
      { organizationResourceId, url },
    )
  }

  options: Object[]

  render = () => (
    <div className="form-container">
      <FormGroup>
        <Select
          id="link-topic-id"
          name="link-topic"
          onChange={this.onSelect}
          options={this.options}
          placeholder="Select a topic"
          value={this.state.selectedOption}
        />
      </FormGroup>
      <FormGroup>
        <Input
          className="link-url test-link-url"
          onChange={this.onInputChange}
          onKeyPress={this.onKeyPress}
          placeholder="Link URL"
          value={this.state.url}
        />
      </FormGroup>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_organization on Organization {
    id
    resourceId

    availableTopics: topics(first: 100) {
      edges {
        node {
          label: name
          value: resourcePath
        }
      }
    }
  }
`)
