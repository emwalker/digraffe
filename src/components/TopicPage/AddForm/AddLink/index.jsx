// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import type { Relay, RepositoryType, TopicType, UserType } from 'components/types'
import upsertLinkMutation from 'mutations/upsertLinkMutation'

/* eslint jsx-a11y/label-has-for: 0 */

type Props = {
  disabled?: boolean,
  relay: Relay,
  topic: TopicType,
  viewer: UserType,
}

type State = {
  url: string,
}

class AddLink extends Component<Props, State> {
  static defaultProps = {
    disabled: true,
  }

  state = {
    url: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createLink()
  }

  get selectedRepo(): ?RepositoryType {
    return this.props.viewer.selectedRepository
  }

  get orgLogin(): ?string {
    const repo = this.selectedRepo
    if (!repo)
      return null
    return repo.organization.login
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const repo = this.selectedRepo
    const repoName = repo ? repo.name : null

    upsertLinkMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        addParentTopicIds: [this.props.topic.id],
        organizationLogin: this.orgLogin,
        repositoryName: repoName,
        url: this.state.url,
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <dl className="form-group">
      <dt>
        <span
          className="tooltipped tooltipped-w"
          aria-label="Add a link to this topic."
        >
          <label htmlFor="create-link-url">Add link</label>
        </span>
      </dt>
      <dd>
        <input
          className="form-control test-link-url input-sm"
          disabled={this.props.disabled}
          id="create-link-url"
          onChange={this.updateUrl}
          onKeyPress={this.onKeyPress}
          placeholder="Url"
          type="url"
          value={this.state.url}
        />
      </dd>
    </dl>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_viewer on User {
    selectedRepository {
      id
      name

      organization {
        login
      }
    }
  }

  fragment AddLink_topic on Topic {
    id
  }
`)
