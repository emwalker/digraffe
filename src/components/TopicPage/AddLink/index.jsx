// @flow
import React, { Component } from 'react'
import { graphql, createFragmentContainer } from 'react-relay'

import createLinkMutation from '../../../mutations/createLinkMutation'
import type { RelayProps } from '../../types'

/* eslint jsx-a11y/label-has-for: 0 */

type State = {
  url: string,
}

class AddLink extends Component<RelayProps, State> {
  state = {
    url: '',
  }

  onKeyPress = (event: Object) => {
    if (event.key === 'Enter')
      this.createLink()
  }

  get relayConfigs() {
    return [{
      type: 'RANGE_ADD',
      parentID: this.props.topic.id,
      connectionInfo: [{
        key: 'Topic_links',
        rangeBehavior: 'prepend',
      }],
      edgeName: 'linkEdge',
    }]
  }

  updateUrl = (event: Object) => {
    this.setState({ url: event.currentTarget.value })
  }

  createLink() {
    const { resourceId: organizationId } = this.props.organization

    createLinkMutation(
      this.props.relay.environment,
      this.relayConfigs,
      {
        organizationId,
        url: this.state.url,
        topicIds: [this.props.topic.resourceId],
      },
    )
    this.setState({ url: '' })
  }

  render = () => (
    <div>
      <dl className="form-group">
        <dt>
          <label htmlFor="create-link-url">Add link</label>
        </dt>
        <dd>
          <input
            className="form-control test-link-url input-sm"
            id="create-link-url"
            onChange={this.updateUrl}
            onKeyPress={this.onKeyPress}
            placeholder="Url"
            value={this.state.url}
          />
        </dd>
      </dl>
    </div>
  )
}

export default createFragmentContainer(AddLink, graphql`
  fragment AddLink_organization on Organization {
    resourceId
  }

  fragment AddLink_topic on Topic {
    id
    resourceId
  }
`)
