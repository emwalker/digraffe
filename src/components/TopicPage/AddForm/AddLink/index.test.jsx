import React from 'react'
import { shallow } from 'enzyme'

import AddLink from './index'

jest.mock('react-relay', () => ({ createFragmentContainer: component => component }))

const props = {
  disabled: false,
  relay: { environment: {} },
  topic: {},
  viewer: {},
}

describe('<AddLink />', () => {
  const wrapper = shallow(<AddLink {...props} />)

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('has a tooltip', () => {
    expect(wrapper.find('.tooltipped').exists()).toBeTruthy()
  })
})
