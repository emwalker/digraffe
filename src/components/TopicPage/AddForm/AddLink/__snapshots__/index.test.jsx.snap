// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddLink /> renders 1`] = `
<dl
  className="form-group"
>
  <dt>
    <span
      aria-label="Add a link to this topic."
      className="tooltipped tooltipped-w"
    >
      <label
        htmlFor="create-link-url"
      >
        Add link
      </label>
    </span>
  </dt>
  <dd>
    <textarea
      className="form-control test-link-url input-sm"
      disabled={false}
      id="create-link-url"
      onChange={[Function]}
      onKeyPress={[Function]}
      placeholder="Url"
      rows={1}
      type="url"
      value=""
    />
  </dd>
</dl>
`;
